internal/controllers/machine/machine_controller.go

```diff
@@ -784,21 +784,25 @@ func (r *Reconciler) isDeleteNodeAllowed(ctx context.Context, cluster *clusterv1
 		}
 	}
 
-	// Get all of the active machines that belong to this cluster.
-	machines, err := collections.GetFilteredMachinesForCluster(ctx, r.Client, cluster, collections.ActiveMachines)
+	// Get all the machines that belong to this cluster.
+	machines, err := collections.GetFilteredMachinesForCluster(ctx, r.Client, cluster)
 	if err != nil {
 		return err
 	}
 
 	// Whether or not it is okay to delete the NodeRef depends on the
 	// number of remaining control plane members and whether or not this
 	// machine is one of them.
-	numControlPlaneMachines := len(machines.Filter(collections.ControlPlaneMachines(cluster.Name)))
-	if numControlPlaneMachines == 0 {
-		// Do not delete the NodeRef if there are no remaining members of
+	controlPlaneMachines := machines.Filter(collections.ControlPlaneMachines(cluster.Name))
+	isMachineAControlPlaneMachine := slices.Contains(controlPlaneMachines.Names(), machine.Name)
+	numOfControlPlaneMachines := len(controlPlaneMachines)
+
+	if isMachineAControlPlaneMachine && numOfControlPlaneMachines == 1 {
+		// Do not delete the NodeRef as this is the last remaining member of
 		// the control plane.
-		return errNoControlPlaneNodes
+		return errLastControlPlaneNode
 	}
+
```

Comment:
What if we do
```
if util.IsControlPlaneMachine(m) {
   // get control plane machines (call GetFilteredMachinesForCluster with collections.ControlPlaneMachines filter)
   // if there is only one and it is m, return error
}
```
So we read machines only when we need to check them, and we also drop workers asap

Reply:
Nice optimization, thanks.